Detectors/EMCAL/workflow/src/RecoWorkflow.cxx

```diff
@@ -0,0 +1,155 @@
+#include "Framework/DataSpecUtils.h"
+#include "SimulationDataFormat/MCCompLabel.h"
+#include "SimulationDataFormat/MCTruthContainer.h"
+
+namespace o2
+{
+
+template <typename T>
+using BranchDefinition = framework::MakeRootTreeWriterSpec::BranchDefinition<T>;
+
+namespace emcal
+{
+
+namespace reco_workflow
+{
+
+const std::unordered_map<std::string, InputType> InputMap{
+  { "digitizer", InputType::Digitizer },
+  { "digits", InputType::Digits },
+  { "raw", InputType::Raw },
+  { "clusters", InputType::Clusters },
+};
+
+const std::unordered_map<std::string, OutputType> OutputMap{
+  { "digits", OutputType::Digits },
+  { "raw", OutputType::Raw },
+  { "clusters", OutputType::Clusters }
+};
+
+o2::framework::WorkflowSpec getWorkflow(bool propagateMC,
+                                        std::string const& cfgInput,
+                                        std::string const& cfgOutput)
+{
+  InputType inputType;
+
+  try {
+    inputType = InputMap.at(cfgInput);
+  } catch (std::out_of_range&) {
+    throw std::invalid_argument(std::string("invalid input type: ") + cfgInput);
+  }
+  std::vector<OutputType> outputTypes;
+  try {
+    outputTypes = RangeTokenizer::tokenize<OutputType>(cfgOutput, [](std::string const& token) { return OutputMap.at(token); });
+  } catch (std::out_of_range&) {
+    throw std::invalid_argument(std::string("invalid output type: ") + cfgOutput);
+  }
+  auto isEnabled = [&outputTypes](OutputType type) {
+    return std::find(outputTypes.begin(), outputTypes.end(), type) != outputTypes.end();
+  };
+
+  o2::framework::WorkflowSpec specs;
+
+  if (inputType == InputType::Digits) {
+    specs.emplace_back(o2::emcal::getPublisherSpec(PublisherConf{
+                                                     "emcal-digit-reader",
+                                                     "o2sim",
+                                                     { "digitbranch", "EMCALDigit", "Digit branch" },
+                                                     { "mcbranch", "EMCALDigitMCTruth", "MC label branch" },
+                                                     o2::framework::OutputSpec{ "EMC", "DIGITS" },
+                                                     o2::framework::OutputSpec{ "EMC", "DIGITSMCTR" } },
+                                                   propagateMC));
+
+    specs.emplace_back(o2::emcal::reco_workflow::getEmcalDigitsPrinterSpec());
+  }
+
+  // check if the process is ready to quit
+  // this is decided upon the meta information in the EMCAL block header, the operation is set
+  // value kNoPayload in case of no data or no operation
+  // see also PublisherSpec.cxx
+  // in this workflow, the EOD is sent after the last real data, and all inputs will receive EOD,
+  // so it is enough to check on the first occurence
+  // FIXME: this will be changed once DPL can propagate control events like EOD
+  auto checkReady = [](o2::framework::DataRef const& ref, bool& isReady) {
```

Comment:
I suggest to remove this completely because this particular case should be solved generically in the framework.

Reply:
For the moment it is needed as other components in particular in the QC need to check this.

Reply:
ok, fair enough, so you want to merge this workflow with another one, but in this case you don't need the writer at all. However, I'm fine with that approach.

There will be a conflict with #2124 because I modified the return value of the evaluator functor.

Reply:
I removed the isReady argument. As local variable it is not needed.